scaletest/dynamicparameters/run.go

```diff
@@ -0,0 +1,114 @@
+package dynamicparameters
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"slices"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/scaletest/harness"
+	"github.com/coder/websocket"
+)
+
+type Runner struct {
+	client *codersdk.Client
+	cfg    Config
+}
+
+var _ harness.Runnable = &Runner{}
+
+func NewRunner(client *codersdk.Client, cfg Config) *Runner {
+	clone := codersdk.New(client.URL)
+	clone.HTTPClient = client.HTTPClient
+	clone.SetLogger(client.Logger())
+	clone.SetSessionToken(cfg.SessionToken)
+	return &Runner{
+		client: clone,
+		cfg:    cfg,
+	}
+}
+
+// Run executes the dynamic parameters test, which:
+//
+// 1. connects to the dynamic parameters stream
+// 2. waits for the initial response
+// 3. sends a change request
+// 4. waits for the change response
+// 5. closes the stream
+func (r *Runner) Run(ctx context.Context, _ string, logs io.Writer) (retErr error) {
+	startTime := time.Now()
+	stream, err := r.client.TemplateVersionDynamicParameters(ctx, codersdk.Me, r.cfg.TemplateVersion)
+	if err != nil {
+		return xerrors.Errorf("connect to dynamic parameters stream: %w", err)
+	}
+	defer stream.Close(websocket.StatusNormalClosure)
+	respCh := stream.Chan()
+
+	var initTime time.Time
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case resp, ok := <-respCh:
+		if !ok {
+			return xerrors.Errorf("dynamic parameters stream closed before initial response")
+		}
+		initTime = time.Now()
+		r.cfg.Metrics.LatencyInitialResponseSeconds.
+			WithLabelValues(r.cfg.MetricLabelValues...).
+			Observe(initTime.Sub(startTime).Seconds())
+		_, _ = fmt.Fprintf(logs, "initial response: %+v\n", resp)
+		if !slices.ContainsFunc(resp.Parameters, func(p codersdk.PreviewParameter) bool {
+			return p.Name == "zero"
+		}) {
+			return xerrors.Errorf("mising expected parameter: 'zero'")
+		}
+		if err := checkNoDiagnostics(resp); err != nil {
+			return xerrors.Errorf("unexpected initial response diagnostics: %w", err)
+		}
+	}
+
+	err = stream.Send(codersdk.DynamicParametersRequest{
+		ID: 1,
+		Inputs: map[string]string{
+			"zero": "B",
+		},
+	})
+	if err != nil {
+		return xerrors.Errorf("send change request: %w", err)
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case resp, ok := <-respCh:
+		if !ok {
+			return xerrors.Errorf("dynamic parameters stream closed before change response")
+		}
+		_, _ = fmt.Fprintf(logs, "change response: %+v\n", resp)
+		r.cfg.Metrics.LatencyChangeResponseSeconds.
+			WithLabelValues(r.cfg.MetricLabelValues...).
+			Observe(time.Since(initTime).Seconds())
+		if resp.ID != 1 {
+			return xerrors.Errorf("unexpected response ID: %d", resp.ID)
+		}
+		if err := checkNoDiagnostics(resp); err != nil {
+			return xerrors.Errorf("unexpected change response diagnostics: %w", err)
+		}
+		return nil
```

Comment:
We do not assert any param condition on the response. I wonder if we can refactor the `ParameterAsserter` structure to be useful here:

https://github.com/coder/coder/blob/0b6d03c28a5507ddd894011139c101d41075c5d0/coderd/coderdtest/dynamicparameters.go#L122-L122

It really streamlines the assertion code to something like:
```golang
	coderdtest.AssertParameter(t, "groups", resp.Parameters).
		Exists().
		Options(database.EveryoneGroup, "admin", "auditor").
		Value("admin")
```

It just currently only works in unit tests.